gadget/install/install.go

```diff
@@ -377,6 +376,141 @@ func Run(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string, options
 	}, nil
 }
 
+// structureFromPartDevice returns the OnDiskStructure for a partition
+// device.
+func structureFromPartDevice(diskVol *gadget.OnDiskVolume, partDevice string) (*gadget.OnDiskStructure, error) {
+	for _, p := range diskVol.Structure {
+		// We remove /dev from p.Node here
+		if filepath.Base(p.Node) == partDevice {
+			return &p, nil
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find partition %q", partDevice)
+}
+
+// laidOutStructureForDiskStructure searches for the laid out structure that
+// matches a given OnDiskStructure.
+func laidOutStructureForDiskStructure(laidVols map[string]*gadget.LaidOutVolume, gadgetVolName string, onDiskStruct *gadget.OnDiskStructure) (
+	*gadget.LaidOutStructure, error) {
+
+	for _, laidVol := range laidVols {
+		// Check that this is the right volume
+		if laidVol.Name != gadgetVolName {
+			continue
+		}
+		for _, laidStruct := range laidVol.LaidOutStructure {
+			if onDiskStruct.Name == laidStruct.Name {
+				return &laidStruct, nil
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("cannot find laid out structure for %q", onDiskStruct.Name)
+}
+
+// sysfsPathForBlockDevice returns the sysfs path for a block device.
+func sysfsPathForBlockDevice(device string) (string, error) {
+	partPath, err := os.Readlink(filepath.Join("/sys/class/block", device))
+	if err != nil {
+		return "", err
```

Comment:
related to the validation comment from before, seems then this error should be wrapped in something like "cannot find..." or "cannot resolve..." ?

Reply:
Done